utilities/milter-regex/milter-regex.8
.\" $Id: milter-regex.8,v 1.1.1.1 2004-10-12 08:19:19 vak Exp $ */
.\"
.\" Copyright (c) 2003-2004 Daniel Hartmeier
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\"    - Redistributions of source code must retain the above copyright
.\"      notice, this list of conditions and the following disclaimer. 
.\"    - Redistributions in binary form must reproduce the above
.\"      copyright notice, this list of conditions and the following
.\"      disclaimer in the documentation and/or other materials provided
.\"      with the distribution. 
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
.\" FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
.\" COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
.\" BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
.\" LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
.\" CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
.\" ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd September 24, 2003
.Dt MILTER-REGEX 8
.Os
.Sh NAME
.Nm milter-regex
.Nd sendmail milter plugin for regular expression filtering
.Sh SYNOPSIS
.Nm
.Op Fl d
.Op Fl c Ar config
.Op Fl p Ar pipe
.Op Fl u Ar user
.Sh DESCRIPTION
The
.Nm
plugin can be used with the milter API of
.Xr sendmail 8
to filter mails using regular expressions matching SMTP envelope
parameters and mail headers and body.
.Pp
The options are as follows:
.Bl -tag -width "-c config"
.It Fl d
Don't detach from controlling terminal and produce verbose debug
output on stdout.
.It Fl c Ar config
Use the specified configuration file instead of the default,
/etc/milter-regex.conf.
.It Fl p Ar pipe 
Use the specified pipe to interface
.Xr sendmail 8 .
Default is unix:/var/spool/milter-regex/sock.
.It Fl u Ar user
Run as the specified user instead of the default, _milter-regex.
When
.Nm
is started as root, it calls
.Xr setuid 2
to drop privileges.
The non-privileged user should have read access to the configuration
file and read-write access to the pipe.
.El
.Pp
.Sh SENDMAIL CONFIGURATION
.Xr sendmail 8
needs to be compiled with milter support.
This can be done with:
.Bd -literal -offset indent
# cd /usr/src/gnu/usr.sbin/sendmail
# make clean cleandir
# setenv WANT_LIBMILTER 1
# make obj && make && make install
.Ed
.Pp
Then, the plugin needs to be registered in the
.Xr sendmail 8
configuration, by adding the following lines to the .mc file
.Bd -literal -offset indent
INPUT_MAIL_FILTER(`milter-regex',
	`S=unix:/var/spool/milter-regex/sock, T=S:30s;R:2m')
.Ed
.Pp
rebuilding /etc/mail/sendmail.cf from the .mc file using
.Xr m4 1 ,
and restarting
.Xr sendmail 8 .
.Pp
.Sh PLUGIN CONFIGURATION
The configuration file consists of rules that, when matched, cause
.Xr sendmail 8
to reject mails.
Emtpy lines and lines starting with # are ignored, as well as
leading whitespace (blanks, tabs).
Trailing backslashes can be used to wrap long rules into multiple
lines.
Each rule starts with one of the following commands:
.Bl -tag -width "xxxx"
.It reject <message>
Subsequent rules cause the mail to be rejected with a permanent
error consisting of the specified text part.
The SMTP reply consists of the three-digit code 550 (RFC 2821
"command rejected for policy reasons"), the extended reply code
5.7.1 (RFC 1893 "Permanent Failure", "Security or Policy Status",
"Delivery not authorized, message refused") and the text part
(which defaults to "Command rejected", if not specified).
This is a permanent failure, which causes the sender to remove the
message from its queue without trying to retransmit, commonly
generating a bounce message to the sender.
.It tempfail <message>
Subsequent matching rules cause the mail to be rejected with
a temporary error consisting of the specified text part.
The SMTP reply consists of the three-digit code 451 (RFC 2821
"Requested action aborted: local error in processing"), the extended
reply code 4.7.1 (RFC 1893 "Persistent Transient Failure",
"Security or Policy Status", "Delivery not authorized, message refused")
and the text part (which defaults to "Please try again later",
if not specified).
This is a temporary failure, which causes the sender to keep the
message in its queue and try to retransmit it, commonly for several
days.
.It discard
Subsequent matching rules cause the mail to be accepted but then
discarded silently.
Note that
.Em connect
and
.Em helo
rules should not use
.Em discard .
.It accept
Subsequent matching rules cause the mail to be accepted without
further rule evaluation.
Can be used for whitelist criteria.
.El
.Pp
A command is followed by one or more expressions, each causing
the previous command to be executed when matched.
The following expressions can be used:
.Bl -tag -width "xxxx"
.It connect <hostname> <address>
Reject the connection if both the sender's hostname and address
match the specified regular expressions.
The numerical address is either dotted-quad (IPv4) or coloned-hex (IPv6).
The hostname is the result of a DNS reverse resolution of the numerical
address (which
.Xr sendmail 8
performs independantly of the milter plugin).
When resolution fails, the hostname contains the numerical
address in square brackets.
.It helo <name>
Reject the connection if the sender supplied HELO name matches
the specified regular expression.
Commonly, the sender supplies his fully-qualified hostname as
HELO name.
.It envfrom <address>
Reject the mail if the sender supplied envelope MAIL FROM
address matches the specified regular expression.
Addresses commonly have the form <user@host.doma.in>.
.It envrcpt <address>
Reject the mail if the sender supplied envelope RCPT TO
address matches the specified regular expression.
.It header <name> <value>
Reject the mail if a header matches the specified name and value.
For instance, the header "Subject: Test" matches name Subject
and value Test.
.It body <line>
Reject the mail if a body line matches the specified regular
expression.
.El
.Pp
The plugin regularly checks the configuration file for modification
and reloads it automatically.
Signals like SIGHUP will terminate the plugin, according
to the milter signal handler.
The plugin reacts to any kind of error, like syntax errors in the
configuration file, by failing open, accepting all messages.
When the plugin is not running,
.Xr sendmail 8
will accept all messages.
.Pp
.Sh REGULAR EXPRESSIONS
The regular expressions used in the configuration rules are
enclosed in arbitrary delimiters, no further escaping is needed.
.Pp
The first character of an argument is taken as the delimiter,
and all subsequent characters up to the next occurance of the
same delimiter are taken literally as the regular expression.
Since the delimiter itself cannot be part of the regular
expression (no escaping is supported), a delimiter must be
chosen that doesn't occur in the regular expression itself.
Each argument can use a different delimiter, all characters
except spaces and tabs are valid.
.Pp
Two immediately adjacent delimiters form an empty regular
expression, which always matches and requires no
.Xr regexec 3
call.
This can be used in rules requiring multiple arguments, to
match only some arguments.
.Pp
See
.Xr re_format 7
for a detailed description of basic and extended regular expressions.
.Pp
Optionally, the following flags can be used after the closing
delimiter:
.Bl -tag -width "xxx" -compact
.It e
Extended regular expression.
This sets REG_EXTENDED for
.Xr regcomp 3 .
.It i
Ignore upper/lower case.
This sets REG_ICASE.
.It n
Not matching.
Reverses the matching result, i.e. the mail is rejected if the
regular expression does not match.
.El
.Sh BOOLEAN EXPRESSIONS
A rule can consist of either a simple term or more complex
expressions.
A term has the form
.Bd -literal
header /From/ /domain/i
.Ed
.Pp
and expressions can be built combining terms with operators
"and", "or", "not" and parentheses, as in
.Bd -literal
header /From/ /domain/i and body /money/
( not header /From/ /domain/ ) and ( body /sex/ or body /fast/ )
.Ed
.Pp
Operator precedence should not be relied on, instead parentheses
should be used to resolve any ambiguities (they usually produce
syntax errors from the parser).
.Sh MACROS
Macros allow to store terms or expressions as a name, and $name
can be used as term within other rules, expressions or macro
definitions.
Example:
.Bd -literal
friends         = header /^Received$/ /^from [^ ]*(ork\.net|home\.com)/e
attachments     = header ,^Content-Type$, ,multipart/mixed, and \\
                    body ,^Content-Type: application/,
executables     = $attachments and body ,name=".*\.(pif|exe|scr)"$,e

reject "executable attachment from non-friends"
$executables and not $friends
.Ed
.Pp
Macro names must begin with a letter and may contain alphanumeric
characters and punctuation characters.
Reserved keywords (like "reject" or "header") cannot be used as
macro names.
Macros must be defined before use, the definition must precede the
use in the configuration file, read from top to bottom.
.Sh EVALUATION
Rules are evaluated in the order specified in the configuration
file, from top to bottom.
When a rule matches, the correpsonding action is taken, that is
the last action specified before the matching rule.
.Pp
The plugin evaluates the rules every time a line of mail (or
envelope) is received.
As soon as a rule matches, the action is taken immediately,
possibly before the entire mail is received, even if further
lines might possibly make other rules match, too.
This means the first rule matching chronologically has precendence.
.Pp
If evaluation for a line of mail makes two (or more) rules match,
the rule that comes first in the configuration file has precendence.
.Pp
Boolean expressions are short-circuit evaluated, that means
"a or b" becomes true as soon as one of the terms is true and
"a and b" becomes false as soon as one of the terms is false,
even if the other term is not known, possibly because the relevant
mail line has not been received yet.
.Sh EXAMPLES
.Bd -literal
# /etc/milter-regex.conf example

tempfail "Sender IP address not resolving"
connect /\\[.*\\]/ //

reject "Malformed HELO (not a domain, no dot)"
helo /\\./n

reject "Malformed RCPT TO (not an email address, not <.*@.*>)"
envrcpt /<(.*@.*|Postmaster)>/ein

reject "HTML mail not accepted"
# use comma as delimiter here, as / occurs within RE
header /^Content-type$/i ,^text/html,i
body ,^Content-type: text/html,i

# Swen worm
discard
header /^(TO|FROM|SUBJECT)$/e //
header /^Content-type$/i /boundary="Boundary_(ID_/i
header /^Content-type$/i /boundary="[a-z]*"/

utilities/milter-regex/milter-regex.c
/* $Id: milter-regex.c,v 1.7 2004-11-11 22:25:32 vak Exp $ */

/*
 * Copyright (c) 2003-2004 Daniel Hartmeier
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *    - Redistributions of source code must retain the above copyright
 *      notice, this list of conditions and the following disclaimer.
 *    - Redistributions in binary form must reproduce the above
 *      copyright notice, this list of conditions and the following
 *      disclaimer in the documentation and/or other materials provided
 *      with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 */

static const char rcsid[] = "$Id: milter-regex.c,v 1.7 2004-11-11 22:25:32 vak Exp $";

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <errno.h>
#include <fcntl.h>
#include <pwd.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <unistd.h>
#include <pthread.h>
#ifdef __linux__
#include <grp.h>
#endif
#include <libmilter/mfapi.h>

#include "eval.h"

extern int parse_ruleset(const char *, struct ruleset **, char *, size_t);

static const char	*rule_file_name = "/etc/milter-regex.conf";
static int		 debug = 0;

struct context {
	struct ruleset	*rs;
	int		*res;
	char		 buf[2048];	/* longer body lines are wrapped */
	unsigned	 pos;		/* write position within buf */
	char		 host[128];
	char		 addr[64];
	char		 hdr_from[128];
	char		 hdr_to[128];
	char		 hdr_subject[128];
};

static sfsistat		 setreply(SMFICTX *, struct context *,
			    const struct action *, char *, char *);
static struct ruleset	*get_ruleset(void);
static sfsistat		 cb_connect(SMFICTX *, char *, _SOCK_ADDR *);
static sfsistat		 cb_helo(SMFICTX *, char *);
static sfsistat		 cb_envfrom(SMFICTX *, char **);
static sfsistat		 cb_envrcpt(SMFICTX *, char **);
static sfsistat		 cb_header(SMFICTX *, char *, char *);
static sfsistat		 cb_eoh(SMFICTX *);
static sfsistat		 cb_body(SMFICTX *, u_char *, size_t);
static sfsistat		 cb_eom(SMFICTX *);
static sfsistat		 cb_abort(SMFICTX *);
static sfsistat		 cb_close(SMFICTX *);
static void		 usage(const char *);
static void		 msg(int, struct context *, const char *, ...);

#define USER		"_milter-regex"
#define OCONN		"unix:/var/spool/milter-regex/sock"
#define RCODE_REJECT	"554"
#define RCODE_TEMPFAIL	"451"
#define XCODE_REJECT	"5.7.1"
#define XCODE_TEMPFAIL	"4.7.1"
#define	MAXRS		16

static const char	*user = USER;

#if defined(__FreeBSD__) && __FreeBSD__ <= 30
/* no sockaddr_in6 in FreeBSD 3.4 */
#undef AF_INET6
#endif

#if __linux__ || __sun__
#define	ST_MTIME st_mtime
extern size_t	 strlcpy(char *, const char *, size_t);
#else
#define	ST_MTIME st_mtimespec
#endif

#ifdef __sun__
int
daemon(int nochdir, int noclose)
{
	pid_t pid;
	int fd;

	if ((pid = fork()) < 0) {
		fprintf(stderr, "fork: %s\n", strerror(errno));
		return (1);
	} else if (pid > 0)
		_exit(0);
	if ((pid = setsid()) == -1) {
		fprintf(stderr, "setsid: %s\n", strerror(errno));
		return (1);
	}
	if ((pid = fork()) < 0) {
		fprintf(stderr, "fork: %s\n", strerror(errno));
		return (1);
	} else if (pid > 0)
		_exit(0);
	if (!nochdir && chdir("/")) {
		fprintf(stderr, "chdir: %s\n", strerror(errno));
		return (1);
	}
	if (!noclose) {
		dup2(fd, fileno(stdout));
		dup2(fd, fileno(stderr));
		dup2(open("/dev/null", O_RDONLY, 0), fileno(stdin));
	}
	return (0);
}
#endif

void
drop_privileges ()
{
	struct passwd *pw;

	if (getuid() != 0)
		return;
	pw = getpwnam(user);
	if (! pw) {
		fprintf(stderr, "getpwnam: %s: %s\n", user,
		    strerror(errno));
		exit (1);
	}
	setgroups(1, &pw->pw_gid);
	if (setegid(pw->pw_gid) || setgid(pw->pw_gid)) {
		fprintf(stderr, "setgid: %s\n", strerror(errno));
		exit (1);
	}
	if (
#if ! ( __linux__ || __sun__ )
	    seteuid(pw->pw_uid) ||
#endif
	    setuid(pw->pw_uid)) {
		fprintf(stderr, "setuid: %s\n", strerror(errno));
		exit (1);
	}
}

static sfsistat
setreply(SMFICTX *ctx, struct context *context, const struct action *action,
	char *command, char *arg)
{
	int priority;
	char *reply;

	switch (action->type) {
	case SMFIS_REJECT:
		priority = LOG_NOTICE;
		reply = "REJECT";
		break;
	case SMFIS_TEMPFAIL:
		priority = LOG_NOTICE;
		reply = "TEMPFAIL";
		break;
	case SMFIS_DISCARD:
		priority = LOG_NOTICE;
		reply = "DISCARD";
		break;
	case SMFIS_CONTINUE:
		priority = LOG_DEBUG;
		reply = "CONTINUE";
		break;
	case SMFIS_ACCEPT:
	default:
		priority = LOG_INFO;
		reply = "ACCEPT";
		break;
	}
	if (! action->msg || ! *action->msg)
		msg(priority, context, arg ? "%s %s %s" : "%s %s",
			reply, command, arg);
	else if (arg)
		msg(priority, context, "%s %s %s: %s",
			reply, command, arg, action->msg);
	else
		msg(priority, context, "%s %s: %s",
			reply, command, action->msg);

	if (action->type == SMFIS_REJECT &&
	    smfi_setreply(ctx, RCODE_REJECT, XCODE_REJECT,
	    (char *)action->msg) != MI_SUCCESS)
		msg(LOG_ERR, context, "smfi_setreply");
	if (action->type == SMFIS_TEMPFAIL &&
	    smfi_setreply(ctx, RCODE_TEMPFAIL, XCODE_TEMPFAIL,
	    (char *)action->msg) != MI_SUCCESS)
		msg(LOG_ERR, context, "smfi_setreply");
	return (action->type);
}

static struct ruleset *
get_ruleset(void)
{
	static struct ruleset *rs[MAXRS] = {};
	static int cur = 0;
	static time_t last_check = 0;
	static struct stat sbo;
	static pthread_mutex_t mutex;
	time_t t = time(NULL);
	int load = 0;

	if (!last_check) {
		drop_privileges ();
		memset(&sbo, 0, sizeof(sbo));
		if (pthread_mutex_init(&mutex, 0)) {
			msg(LOG_ERR, NULL, "parse_ruleset: pthread_mutex_init failed");
			return NULL;
		}
	}
	if (pthread_mutex_lock(&mutex))
		msg(LOG_ERR, NULL, "parse_ruleset: pthread_mutex_lock failed");
	if (t - last_check >= 10) {
		struct stat sb;

		last_check = t;
		memset(&sb, 0, sizeof(sb));
		if (stat(rule_file_name, &sb))
			msg(LOG_ERR, NULL, "get_ruleset: stat: %s: %s",
			    rule_file_name, strerror(errno));
		else if (memcmp(&sb.ST_MTIME, &sbo.ST_MTIME,
		    sizeof(sb.ST_MTIME))) {
			memcpy(&sbo.ST_MTIME, &sb.ST_MTIME,
			    sizeof(sb.ST_MTIME));
			load = 1;
		}
	}
	if (load || rs[cur] == NULL) {
		int i;
		char err[8192];

		msg(LOG_DEBUG, NULL, "loading new configuration file");
		for (i = 0; i < MAXRS; ++i)
			if (rs[i] != NULL && rs[i]->refcnt == 0) {
				msg(LOG_DEBUG, NULL, "freeing unused ruleset "
				    "%d/%d", i, MAXRS);
				free_ruleset(rs[i]);
				rs[i] = NULL;
			}
		for (i = 0; i < MAXRS; ++i)
			if (rs[i] == NULL)
				break;
		if (i == MAXRS)
			msg(LOG_ERR, NULL, "all rulesets are in use, cannot "
			    "load new one", MAXRS);
		else if (parse_ruleset(rule_file_name, &rs[i], err,
		    sizeof(err)) || rs[i] == NULL)
			msg(LOG_ERR, NULL, "parse_ruleset: %s", err);
		else {
			msg(LOG_INFO, NULL, "configuration file %s loaded "
			    "successfully", rule_file_name);
			cur = i;
		}
	}
	if (pthread_mutex_unlock(&mutex))
		msg(LOG_ERR, NULL, "parse_ruleset: pthread_mutex_unlock failed");
	return (rs[cur]);
}

static sfsistat
cb_connect(SMFICTX *ctx, char *name, _SOCK_ADDR *sa)
{
	struct context *context;
	const struct action *action;
	char host[64];

	context = calloc(1, sizeof(*context));
	if (context == NULL) {
		msg(LOG_ERR, NULL, "cb_connect: calloc: %s", strerror(errno));
		return (SMFIS_ACCEPT);
	}
	context->rs = get_ruleset();
	if (context->rs == NULL) {
		free(context);
		msg(LOG_ERR, NULL, "cb_connect: get_ruleset");
		return (SMFIS_ACCEPT);
	}
	context->res = calloc(context->rs->maxidx, sizeof(*context->res));
	if (context->res == NULL) {
		free(context);
		msg(LOG_ERR, NULL, "cb_connect: calloc: %s", strerror(errno));
		return (SMFIS_ACCEPT);
	}
	if (smfi_setpriv(ctx, context) != MI_SUCCESS) {
		free(context->res);
		free(context);
		msg(LOG_ERR, NULL, "cb_connect: smfi_setpriv");
		return (SMFIS_ACCEPT);
	}
	context->rs->refcnt++;

	strlcpy(host, "unknown", sizeof(host));
	switch (sa ? sa->sa_family : 0) {
	case AF_INET: {
		struct sockaddr_in *sin = (struct sockaddr_in *)sa;

		if (inet_ntop(AF_INET, &sin->sin_addr.s_addr, host,
		    sizeof(host)) == NULL)
			msg(LOG_ERR, NULL, "cb_connect: inet_ntop: %s",
			    strerror(errno));
		break;
	}
#ifdef AF_INET6
	case AF_INET6: {
		struct sockaddr_in6 *sin6 = (struct sockaddr_in6 *)sa;

		if (inet_ntop(AF_INET6, &sin6->sin6_addr, host,
		    sizeof(host)) == NULL)
			msg(LOG_ERR, NULL, "cb_connect: inet_ntop: %s",
			    strerror(errno));
		break;
	}
#endif
	}
	strlcpy(context->host, name, sizeof(context->host));
	strlcpy(context->addr, host, sizeof(context->addr));
	msg(LOG_DEBUG, context, "cb_connect('%s', '%s')", name, host);
	if ((action = eval_cond(context->rs, context->res, COND_CONNECT,
	    name, host)) != NULL)
		return (setreply(ctx, context, action, "CONNECT", 0));
	if ((action = eval_end(context->rs, context->res, COND_CONNECT)) !=
	    NULL)
		return (setreply(ctx, context, action, "CONNECT", 0));
	return (SMFIS_CONTINUE);
}

static sfsistat
cb_helo(SMFICTX *ctx, char *arg)
{
	struct context *context;
	const struct action *action;

	if ((context = (struct context *)smfi_getpriv(ctx)) == NULL) {
		msg(LOG_ERR, NULL, "cb_helo: smfi_getpriv");
		return (SMFIS_ACCEPT);
	}
	msg(LOG_DEBUG, context, "cb_helo('%s')", arg);
	/* multiple HELO imply RSET in sendmail */
	eval_clear(context->rs, context->res, COND_HELO);
	if ((action = eval_cond(context->rs, context->res, COND_HELO,
	    arg, NULL)) != NULL)
		return (setreply(ctx, context, action, "HELO", arg));
	if ((action = eval_end(context->rs, context->res, COND_HELO)) !=
	    NULL)
		return (setreply(ctx, context, action, "HELO", arg));
	return (SMFIS_CONTINUE);
}

static sfsistat
cb_envfrom(SMFICTX *ctx, char **args)
{
	struct context *context;
	const struct action *action;

	if ((context = (struct context *)smfi_getpriv(ctx)) == NULL) {
		msg(LOG_ERR, NULL, "cb_envfrom: smfi_getpriv");
		return (SMFIS_ACCEPT);
	}
	/* multiple MAIL FROM indicate separate messages */
	eval_clear(context->rs, context->res, COND_ENVFROM);
	context->hdr_subject[0] = 0;
	if (*args != NULL) {
		msg(LOG_DEBUG, context, "cb_envfrom('%s')", *args);
		if ((action = eval_cond(context->rs, context->res, COND_ENVFROM,
		    *args, NULL)) != NULL)
			return (setreply(ctx, context, action, "ENVFROM",
				*args));
	}
	if ((action = eval_end(context->rs, context->res, COND_ENVFROM)) !=
	    NULL)
		return (setreply(ctx, context, action, "ENVFROM", *args));
	return (SMFIS_CONTINUE);
}

static sfsistat
cb_envrcpt(SMFICTX *ctx, char **args)
{
	struct context *context;
	const struct action *action;

	if ((context = (struct context *)smfi_getpriv(ctx)) == NULL) {
		msg(LOG_ERR, NULL, "cb_envrcpt: smfi_getpriv");
		return (SMFIS_ACCEPT);
	}
	if (*args != NULL) {
		msg(LOG_DEBUG, context, "cb_envrcpt('%s')", *args);
		if ((action = eval_cond(context->rs, context->res, COND_ENVRCPT,
		    *args, NULL)) != NULL)
			return (setreply(ctx, context, action, "ENVRCPT",
				*args));
	}
	return (SMFIS_CONTINUE);
}

static sfsistat
cb_header(SMFICTX *ctx, char *name, char *value)
{
	struct context *context;
	const struct action *action;

	if ((context = (struct context *)smfi_getpriv(ctx)) == NULL) {
		msg(LOG_ERR, context, "cb_header: smfi_getpriv");
		return (SMFIS_ACCEPT);
	}
	if ((action = eval_end(context->rs, context->res, COND_ENVRCPT)) !=
	    NULL)
		return (setreply(ctx, context, action, "ENVRCPT", 0));
	msg(LOG_DEBUG, context, "cb_header('%s', '%s')", name, value);
	if (!strcasecmp(name, "From"))
		strlcpy(context->hdr_from, value, sizeof(context->hdr_from));
	else if (!strcasecmp(name, "To"))
		strlcpy(context->hdr_to, value, sizeof(context->hdr_to));
	else if (!strcasecmp(name, "Subject"))
		strlcpy(context->hdr_subject, value,
		    sizeof(context->hdr_subject));
	if ((action = eval_cond(context->rs, context->res, COND_HEADER,
	    name, value)) != NULL)
		return (setreply(ctx, context, action, name, value));
	return (SMFIS_CONTINUE);
}

static sfsistat
cb_eoh(SMFICTX *ctx)
{
	struct context *context;
	const struct action *action;

	if ((context = (struct context *)smfi_getpriv(ctx)) == NULL) {
		msg(LOG_ERR, NULL, "cb_eoh: smfi_getpriv");
		return (SMFIS_ACCEPT);
	}
	msg(LOG_DEBUG, context, "cb_eoh()");
	memset(context->buf, 0, sizeof(context->buf));
	context->pos = 0;
	if ((action = eval_end(context->rs, context->res, COND_HEADER)) !=
	    NULL)
		return (setreply(ctx, context, action, "HEADER", 0));
	return (SMFIS_CONTINUE);
}

static sfsistat
cb_body(SMFICTX *ctx, u_char *chunk, size_t size)
{
	struct context *context;

	if ((context = (struct context *)smfi_getpriv(ctx)) == NULL) {
		msg(LOG_ERR, NULL, "cb_body: smfi_getpriv");
		return (SMFIS_ACCEPT);
	}
	for (; size > 0; size--, chunk++) {
		context->buf[context->pos] = *chunk;
		if (context->buf[context->pos] == '\n' ||
		    context->pos == sizeof(context->buf) - 1) {
			const struct action *action;

			if (context->pos > 0 &&
			    context->buf[context->pos - 1] == '\r')
				context->buf[context->pos - 1] = 0;
			else
				context->buf[context->pos] = 0;
			context->pos = 0;
			msg(LOG_DEBUG, context, "cb_body('%s')", context->buf);
			if ((action = eval_cond(context->rs, context->res,
			    COND_BODY, context->buf, NULL)) != NULL)
				return (setreply(ctx, context, action,
					"BODY", 0));
		} else
			context->pos++;
	}
	return (SMFIS_CONTINUE);
}

static sfsistat
cb_eom(SMFICTX *ctx)
{
	struct context *context;
	const struct action *action;

	if ((context = (struct context *)smfi_getpriv(ctx)) == NULL) {
		msg(LOG_ERR, NULL, "cb_eom: smfi_getpriv");
		return (SMFIS_ACCEPT);
	}
	msg(LOG_DEBUG, context, "cb_eom()");
	if ((action = eval_end(context->rs, context->res, COND_BODY)) !=
	    NULL)
		return (setreply(ctx, context, action, "BODY", 0));
	else
		msg(LOG_DEBUG, context, "ACCEPT, From: %s, To: %s, "
		    "Subject: %s", context->hdr_from, context->hdr_to,
		    context->hdr_subject);
	return (SMFIS_CONTINUE);
}

static sfsistat
cb_abort(SMFICTX *ctx)
{
	struct context *context;

	if ((context = (struct context *)smfi_getpriv(ctx)) == NULL) {
		msg(LOG_ERR, NULL, "cb_abort: smfi_getpriv");
		return (SMFIS_ACCEPT);
	}
	msg(LOG_DEBUG, context, "cb_abort()");
	/* a RSET doesn't clear HELO in sendmail, but MAIL FROM */
	eval_clear(context->rs, context->res, COND_ENVFROM);
	return (SMFIS_CONTINUE);
}

static sfsistat
cb_close(SMFICTX *ctx)
{
	struct context *context;

	context = (struct context *)smfi_getpriv(ctx);
	msg(LOG_DEBUG, context, "cb_close()");
	if (context != NULL) {
		smfi_setpriv(ctx, NULL);
		free(context->res);
		context->rs->refcnt--;
		free(context);
	}
	return (SMFIS_CONTINUE);
}

struct smfiDesc smfilter = {
	"milter-regex",	/* filter name */
	SMFI_VERSION,	/* version code -- do not change */
	SMFIF_ADDHDRS|SMFIF_CHGHDRS|SMFIF_ADDRCPT|SMFIF_DELRCPT, /* flags */
	cb_connect,	/* connection info filter */
	cb_helo,	/* SMTP HELO command filter */
	cb_envfrom,	/* envelope sender filter */
	cb_envrcpt,	/* envelope recipient filter */
	cb_header,	/* header filter */
	cb_eoh,		/* end of header */
	cb_body,	/* body block */
	cb_eom,		/* end of message */
	cb_abort,	/* message aborted */
	cb_close	/* connection cleanup */
};

static void
msg(int priority, struct context *context, const char *fmt, ...)
{
	va_list ap;
	char msg[8192];

	if (priority == LOG_DEBUG && debug == 0)
		return;
	va_start(ap, fmt);
	if (context != NULL)
		snprintf(msg, sizeof(msg), "%s: ", context->host);
	else
		msg[0] = 0;
	vsnprintf(msg + strlen(msg), sizeof(msg) - strlen(msg), fmt, ap);
	if (debug > 1)
		printf("syslog: %s\n", msg);
	else
		syslog(priority, "%s", msg);
	va_end(ap);
}

static void
usage(const char *argv0)
{
	fprintf(stderr, "usage: %s [-d] [-c config] [-u user] "
	    "[-p pipe]\n", argv0);
	exit(1);
}

void
die(const char *reason)
{
	msg(LOG_ERR, NULL, "die: %s", reason);
	smfi_stop();
	sleep(60);
	/* not reached, smfi_stop() kills thread */
	abort();
}

int
main(int argc, char **argv)
{
	int ch;
	const char *oconn = OCONN;
	sfsistat r = MI_FAILURE;
	const char *ofile = NULL;

	tzset();
	openlog("milter-regex", LOG_PID | LOG_NDELAY, LOG_MAIL);

	while ((ch = getopt(argc, argv, "c:dp:u:")) != -1) {
		switch (ch) {
		case 'c':
			rule_file_name = optarg;
			break;
		case 'd':
			++debug;
			break;
		case 'p':
			oconn = optarg;
			break;
		case 'u':
			user = optarg;
			break;
		default:
			usage(argv[0]);
		}
	}
	if (argc != optind) {
		fprintf(stderr, "unknown command line argument: %s ...",
		    argv[optind]);
		usage(argv[0]);
	}

	if (!strncmp(oconn, "unix:", 5))
		ofile = oconn + 5;
	else if (!strncmp(oconn, "local:", 6))
		ofile = oconn + 6;
	if (ofile != NULL)
		unlink(ofile);

	if (smfi_setconn((char *)oconn) != MI_SUCCESS) {
		fprintf(stderr, "smfi_setconn: %s: failed\n", oconn);
		goto done;
	}

	if (smfi_register(smfilter) != MI_SUCCESS) {
		fprintf(stderr, "smfi_register: failed\n");
		goto done;
	}

	if (eval_init(SMFIS_ACCEPT)) {
		fprintf(stderr, "eval_init: failed\n");
		goto done;
	}

	/* daemonize (detach from controlling terminal) */
	if (debug < 2 && daemon(0, 0)) {
		fprintf(stderr, "daemon: %s\n", strerror(errno));
		goto done;
	}
	umask(0177);

	msg(LOG_INFO, NULL, "started: %s", rcsid);
	r = smfi_main();
	if (r != MI_SUCCESS)
		msg(LOG_ERR, NULL, "smfi_main: terminating due to error");
	else
		msg(LOG_INFO, NULL, "smfi_main: terminating without error");

done:
	return (r);
}
